core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt

```diff
@@ -0,0 +1,199 @@
+      setOnToggledFullscreen { fullscreen ->
+        setWindowVisibility(
+          if (fullscreen) View.SYSTEM_UI_FLAG_LOW_PROFILE else View.SYSTEM_UI_FLAG_VISIBLE
+        )
+      }
+    }
+  }
+
+  override fun performLongClick(): Boolean {
+    val result = hitTestResult
+    if (result.type == HitTestResult.SRC_ANCHOR_TYPE) {
+      callback.webViewLongClick(result.extra)
+      return true
+    }
+    return super.performLongClick()
+  }
+
+  override fun onCreateContextMenu(menu: ContextMenu) {
+    super.onCreateContextMenu(menu)
+    val result = hitTestResult
+    if (result.type == HitTestResult.IMAGE_ANCHOR_TYPE ||
+      result.type == HitTestResult.IMAGE_TYPE ||
+      result.type == HitTestResult.SRC_IMAGE_ANCHOR_TYPE
+    ) {
+      val saveMenu =
+        menu.add(0, 1, 0, resources.getString(R.string.save_media))
+      saveMenu.setOnMenuItemClickListener {
+        val msg = SaveHandler(zimReaderContainer).obtainMessage()
+        requestFocusNodeHref(msg)
+        true
+      }
+    }
+  }
+
+  override fun onAttachedToWindow() {
+    super.onAttachedToWindow()
+    compositeDisposable.add(
+      sharedPreferenceUtil.textZooms
+        .subscribe {
+          settings.textZoom = it
+        }
+    )
+  }
+
+  override fun onDetachedFromWindow() {
+    super.onDetachedFromWindow()
+    compositeDisposable.clear()
+  }
+
+  override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
+    super.onScrollChanged(l, t, oldl, oldt)
+    val windowHeight = if (measuredHeight > 0) measuredHeight else 1
+    val pages = contentHeight / windowHeight
+    val page = t / windowHeight
+    callback.webViewPageChanged(page, pages)
+  }
+
+  internal class SaveHandler(private val zimReaderContainer: ZimReaderContainer) :
+    Handler() {
+    private fun getDecodedFileName(url: String?, src: String?): String =
+      url?.substringAfterLast("/", "")
+        ?.takeIf { it.contains(".") }
+        ?: src?.substringAfterLast("/", "")
+          ?.substringAfterLast("%3A") ?: ""
+
+    override fun handleMessage(msg: Message) {
+      val url = msg.data["url"] as? String
+      val src = msg.data["src"] as? String
+      if (url != null || src != null) {
+        val fileName = getDecodedFileName(url, src)
+        val dotIndex = fileName.lastIndexOf('.')
+        var root =
+          Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
+        if (instance.externalMediaDirs.isNotEmpty()) {
+          root = instance.externalMediaDirs[0]
+        }
+        var storageDir = File(root, fileName)
+        var newUrl = fileName
+        var i = 2
+        while (storageDir.exists()) {
+          newUrl = fileName.substring(0, dotIndex) + "_" + i + fileName.substring(dotIndex)
+          storageDir = File(root, newUrl)
+          i++
+        }
+        val source = Uri.parse(src)
+        try {
+          zimReaderContainer.load("$source").data.use {
+            it.copyTo(storageDir.outputStream())
```

Comment:
the output stream will also need to be `use`d

Reply:
but in the documentation, it doesn't mention for releasing resources.

Reply:
![image](https://user-images.githubusercontent.com/3358282/92581799-074a3800-f288-11ea-8649-5401abb7e2eb.png)

It does

Reply:
ooh! I thought i am using first `use`  for this. Thanks!